network/src/protocols/ping.rs

```diff
@@ -46,26 +47,35 @@ impl PingHandler {
         network_state: Arc<NetworkState>,
     ) -> (PingHandler, Sender<()>) {
         let (control_sender, control_receiver) = channel(CONTROL_CHANNEL_BUFFER_SIZE);
-
+        let now = Instant::now();
         (
             PingHandler {
                 interval,
                 timeout,
                 connected_session_ids: Default::default(),
                 network_state,
                 control_receiver,
+                unix_epoch: now
+                    .checked_sub(
+                        SystemTime::now()
+                            .duration_since(UNIX_EPOCH)
+                            .expect("Convert system time fail"),
+                    )
+                    .unwrap_or(now),
             },
             control_sender,
         )
     }
 
-    fn ping_received(&self, id: SessionId) {
+    fn ping_received(&mut self, id: SessionId) {
         trace!("received ping from: {:?}", id);
-        self.mark_time(id, None);
+        self.mark_time(id, None, Instant::now());
     }
 
     fn ping_peers(&mut self, context: &ProtocolContext) {
-        let now = SystemTime::now();
+        let mut now = None;
```

Comment:
This really looks ugly. What we want to do here is cache the time, so we can do it straightforwardly. 

* Add a field `now: Instant` in `PingHandler`
* Bump `now` to current time only in `ping_peers` and `received`
* Use `self.now` to get the current time.

Reply:
This is because if there is no task sent, there is no need to get the time here

The update method you said can't do this

Reply:
It's a tradeoff. In `received`, you already called `Instant::now()`, and in `ping_peers`, it has a very high possibility that you will call `Instance::now()`. Also `ping_peers` is only called once every 2  minutes.

Reply:
Have you tried set `let now = Instance::now()` in `ping_peers` directly. I think the performance difference is trivial but the code will be much more simpler.

Reply:
The original logic is to take it directly. It is indeed simpler. Let me change it back. It is not worthwhile to trade complex logic for almost invisible improvement.

Reply:
There's also a merge conflict.